components/ws-manager-mk2/controllers/workspace_controller_test.go

```diff
@@ -24,110 +27,373 @@ import (
-			podLookupKey := types.NamespacedName{Name: "ws-" + WorkspaceName, Namespace: WorkspaceNamespace}
-			createdPod := &corev1.Pod{}
+			// Controller should detect container exit and add Failed condition.
+			expectConditionEventually(ws, string(workspacev1.WorkspaceConditionFailed), string(metav1.ConditionTrue), "")
 
-			// We'll need to retry getting this newly created CronJob, given that creation may not immediately happen.
-			Eventually(func() bool {
-				err := k8sClient.Get(ctx, podLookupKey, createdPod)
-				return err == nil
-			}, timeout, interval).Should(BeTrue())
+			expectFinalizerAndMarkBackupCompleted(ws, pod)
 
-			By("updating the pod starts value")
-			createdWS := &workspacev1.Workspace{}
-			Eventually(func() (int, error) {
-				err := k8sClient.Get(ctx, types.NamespacedName{Name: WorkspaceName, Namespace: WorkspaceNamespace}, createdWS)
+			expectWorkspaceCleanup(ws, pod)
+		})
+	})
+
+	Context("with headless workspaces", func() {
+		var (
+			ws  *workspacev1.Workspace
+			pod *corev1.Pod
+		)
+		BeforeEach(func() {
+			Skip("TODO(wv): Enable once headless workspaces are implemented")
+
+			ws = newWorkspace(uuid.NewString(), "default")
+			ws.Spec.Type = workspacev1.WorkspaceTypePrebuild
+			pod = createWorkspaceExpectPod(ws)
+
+			// Expect headless status to be reported.
+			Eventually(func() (bool, error) {
+				err := k8sClient.Get(ctx, types.NamespacedName{Name: ws.Name, Namespace: ws.Namespace}, ws)
 				if err != nil {
-					return 0, err
+					return false, err
 				}
+				return ws.Status.Headless, nil
+			}, timeout, interval).Should(BeTrue())
 
-				return createdWS.Status.PodStarts, nil
-			}, timeout, interval).Should(Equal(1))
+			// Expect runtime status also gets reported for headless workspaces.
+			expectRuntimeStatus(ws, pod)
+		})
 
-			By("Creating the pod only once")
-			// TODO(cw): remove this hell of a hack once we're on PVC and no longer need to rely on concurrent init processes.
-			// Here we assume that the controller will have deleted the pod because it failed,
-			// and we removed the finalizer.
-			createdPod.Finalizers = []string{}
-			Expect(k8sClient.Update(ctx, createdPod)).To(Succeed())
-			Expect(k8sClient.Delete(ctx, createdPod)).To(Succeed())
-			Eventually(func() bool {
-				err := k8sClient.Get(ctx, podLookupKey, createdPod)
-				if err != nil && errors.IsNotFound(err) {
-					// We have an error and assume we did not find the pod. This is what we want.
-					return true
-				}
+		It("should cleanup on successful exit", func() {
+			// Manually update pod phase & delete pod to simulate pod exiting successfully.
+			updateObjWithRetries(pod, true, func(p *corev1.Pod) {
+				p.Status.Phase = corev1.PodSucceeded
+			})
+			Expect(k8sClient.Delete(ctx, pod)).To(Succeed())
 
-				fmt.Println(createdPod.ResourceVersion)
-				return false
-			}, timeout, interval).Should(BeTrue(), "pod did not go away")
+			// No backup for headless workspace.
+			expectWorkspaceCleanup(ws, pod)
```

Comment:
Comment, non-blocking:
`No backup for headless workspace.` is true for imagebuild. The product of a prebuild is an archive that gets persisted to object storage. I would expect when we implement prebuilds for mk2, that we'd check for a Condition that represents the archive being saved. Whether we call it BackupComplete (like Regular Workspaces) or something else can be determined later.

Reply:
Added some todo comments as reminders

Reply:
You rock, thanks @WVerlaek ! Marking as resolved.